[JIT] Support for registering C++ lambdas as methods on custom C++ class

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32658 [JIT] Fix stateful lambda stuff and simplify code in custom C++ binding API
* #32604 [JIT] pickle serialization for custom bound classes
* #32586 [JIT] Include custom_class.h in torch/script.h
* #32555 [JIT] Serialize attributes and types in ClassType serialization
* **#32553 [JIT] Support for registering C++ lambdas as methods on custom C++ class**

Differential Revision: [D19543269](https://our.internmc.facebook.com/intern/diff/D19543269)

Commit:
[JIT] Support for registering C++ lambdas as methods on custom C++ class

[ghstack-poisoned]

Commit:
Update on "[JIT] Support for registering C++ lambdas as methods on custom C++ class"


Differential Revision: [D19543269](https://our.internmc.facebook.com/intern/diff/D19543269)

[ghstack-poisoned]

zdevito reviewed (commented):
Maybe find someone to review that understands the template stuff. Otherwise seems fine to me.

Commit:
Update on "[JIT] Support for registering C++ lambdas as methods on custom C++ class"


Differential Revision: [D19543269](https://our.internmc.facebook.com/intern/diff/D19543269)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit a31cceba:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=d2f66083c5b64e64355eb0fc580cfa4230bd17d8&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base d2f66083 since Jan 23
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=a31ccebaec691905a0e057deb7d7ac6f069012fc).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_macos_10_13_py3_test](https://dr.pytorch.org/build-details.html?build_id=90124571) from Jan 23



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[JIT] Support for registering C++ lambdas as methods on custom C++ class"


Differential Revision: [D19543269](https://our.internmc.facebook.com/intern/diff/D19543269)

[ghstack-poisoned]

Commit:
Update on "[JIT] Support for registering C++ lambdas as methods on custom C++ class"


Differential Revision: [D19543269](https://our.internmc.facebook.com/intern/diff/D19543269)

[ghstack-poisoned]

smessmer commented on torch/custom_class.h:
```diff
@@ -101,9 +102,22 @@ class class_ {
+          !c10::guts::is_stateless_lambda<std::decay_t<Method>>::value,
+          bool> = false>
+  class_& def(std::string name, Method m) {
+    auto res = def_(name, m, detail::args_t<decltype(m)>{});
```
you wanna `std::move(name)` and `std::move(m)`

smessmer commented on torch/custom_class.h:
```diff
@@ -101,9 +102,22 @@ class class_ {
+          bool> = false>
   class_& def(std::string name, Func f) {
-    auto res = def_(name, f, detail::args_t<decltype(f)>{});
+    auto res = def_(name, f, detail::args_t<decltype(&Func::operator())>{});
```
also move here. Also, why does this need a special case? It should be possible to write this and the !is_stateless_lambda in one implementation.

smessmer commented on torch/custom_class.h:
```diff
@@ -145,14 +159,41 @@ class class_ {
     auto func = [f](c10::intrusive_ptr<CurClass> cur, Types... args) {
       return at::guts::invoke(f, *cur, args...);
     };
     defineMethod<R>(name, std::move(func));
```
move

smessmer commented on torch/custom_class.h:
```diff
@@ -145,14 +159,41 @@ class class_ {
+          !c10::guts::is_stateless_lambda<std::decay_t<Func>>::value,
+          bool> = false>
   class_& def_(std::string name, Func f, detail::types<R, Types...> funcInfo) {
     auto func = [f](c10::intrusive_ptr<CurClass> cur, Types... args) {
```
`[f = std::move(f)]`

smessmer commented on torch/custom_class.h:
```diff
@@ -145,14 +159,41 @@ class class_ {
+          bool> = false>
+  class_& def_(std::string name, Func f, detail::types<R, Types...> funcInfo) {
+    assert_self_type(funcInfo);
+    defineMethod<R>(name, f);
```
move name and f

Commit:
Update on "[JIT] Support for registering C++ lambdas as methods on custom C++ class"


Differential Revision: [D19543269](https://our.internmc.facebook.com/intern/diff/D19543269)

[ghstack-poisoned]

Commit:
Update on "[JIT] Support for registering C++ lambdas as methods on custom C++ class"


Differential Revision: [D19543269](https://our.internmc.facebook.com/intern/diff/D19543269)

[ghstack-poisoned]

Commit:
Update on "[JIT] Support for registering C++ lambdas as methods on custom C++ class"


Differential Revision: [D19543269](https://our.internmc.facebook.com/intern/diff/D19543269)

[ghstack-poisoned]

jamesr66a commented:
@smessmer I've simplified the member function pointer simplification in https://github.com/pytorch/pytorch/pull/32658

smessmer commented on torch/custom_class.h:
```diff
@@ -101,9 +102,23 @@ class class_ {
+      std::enable_if_t<
+          !std::is_member_function_pointer<std::decay_t<Func>>::value,
+          bool> = false>
   class_& def(std::string name, Func f) {
```
You should probably use perfect forwarding instead of moving for `f` so that callers can pass in const references if they want to. Sorry, I might have given wrong advice here earlier.

smessmer commented on torch/custom_class.h:
```diff
@@ -145,12 +160,40 @@ class class_ {
+      std::enable_if_t<
+          std::is_member_function_pointer<std::decay_t<Func>>::value,
+          bool> = false>
   class_& def_(std::string name, Func f, detail::types<R, Types...> funcInfo) {
```
same here

smessmer commented on torch/custom_class.h:
```diff
@@ -145,12 +160,40 @@ class class_ {
+      std::enable_if_t<
+          !std::is_member_function_pointer<std::decay_t<Func>>::value,
+          bool> = false>
+  class_& def_(std::string name, Func f, detail::types<R, Types...> funcInfo) {
```
and here

smessmer reviewed (approved):
generally looks better, but let's do another offline review and see if we can simplify this a bit more before you land it.

Commit:
Update on "[JIT] Support for registering C++ lambdas as methods on custom C++ class"


Differential Revision: [D19543269](https://our.internmc.facebook.com/intern/diff/D19543269)

[ghstack-poisoned]

Commit:
Update on "[JIT] Support for registering C++ lambdas as methods on custom C++ class"


Differential Revision: [D19543269](https://our.internmc.facebook.com/intern/diff/D19543269)

[ghstack-poisoned]

Commit:
Update on "[JIT] Support for registering C++ lambdas as methods on custom C++ class"


Differential Revision: [D19543269](https://our.internmc.facebook.com/intern/diff/D19543269)

[ghstack-poisoned]
